File: 1.Arrays/Q2/Q2.java
import java.util.Arrays;
import java.util.ArrayList;

/**
 * Given an array, 
 * rotate the array to the right by k steps, 
 * where k is non-negative.
 */

public class Q2 {
    public static void main(String[] args) {
        ArrayList<Integer> arrayList = new ArrayList<>(Arrays.asList(1, 2, 3, 4, 5));
        int[] array = {1, 2, 3, 4, 5};
        int k = 3;

        System.out.println(rotateArrayMethod1(arrayList, k));
        System.out.println(Arrays.toString(rotateArrayMethod2(array, k)));
    }

    static ArrayList<Integer> rotateArrayMethod1(ArrayList<Integer> array, int k) {
        if (array.size() == k || array.size() == 0) {
            return array;
        }

        int counter = 1;
        for (int i = 0; i < array.size(); i++) {
            if (counter <= k) {
                int lastIndex = array.size() - 1;
                int lastItem = array.get(lastIndex);
                array.remove(lastIndex);
                array.add(0, lastItem);
                counter++;
            }
        }

        return array;
    }

    static int[] rotateArrayMethod2(int[] array, int k) {
        if (array.length == k || array.length == 0) {
            return array;
        }
        
        k = k % array.length;
        int tempArray[] = new int[k];

        // store last k elements in temp
        for (int i = 0; i < k; i++) {
            tempArray[i] = array[array.length - k + i];
        }

        // shift rest of array right by k
        for (int i = array.length - k - 1; i >= 0; i--) {
            array[i + k] = array[i];
        }

        // copy temp element to front
        for (int i = 0; i < k; i++) {
            array[i] = tempArray[i];
        }

        return array;
    }
}

File: 1.Arrays/Q2/Q2.go
package main

import (
	"fmt"
)

// Given an array, 
// rotate the array to the right by k steps, 
// where k is non-negative.

func rotateArrayMethod1(arr []int, k int) []int {
	if len(arr) == k || len(arr) == 0 {
		return arr
	}

	counter := 1
	for i := 0; i < len(arr); i++ {
		if counter <= k {
			lastItem := arr[len(arr) - 1]
			arr = arr[:len(arr) - 1]
			arr = append([]int{lastItem}, arr...)
			counter++
		}
	}

	return arr
}

func main() {
	var array = []int{1, 2, 3, 4, 5}
	var k = 3

	var newArrayMethod1 = rotateArrayMethod1(array, k)
	fmt.Println(newArrayMethod1)
}